lib/asyncSender.ts

```diff
@@ -0,0 +1,181 @@
+    // retry logic for sending messages on failures hence they can provide their error handlers
+    // for sender_error and session_error.
+    // If the user did not provide their error handler for sender_error and session_error, then we
+    // add our handlers and make sure we clear the timer and reject the promise for sending
+    // messages with appropriate Error.
+    if (!options.onError) {
+      this.on(SenderEvents.senderError, (context: EventContext) => {
+        onError(SenderEvents.senderError, context.sender!.error as Error);
+      });
+    }
+    if (!options.onSessionError) {
+      this.session.on(SessionEvents.sessionError, (context: EventContext) => {
+        onError(SessionEvents.sessionError, context.sender!.error as Error);
+      });
+    }
+  }
+
+  /**
+   * Sends the message on which one can await to ensure that the message has been successfully
+   * delivered.
+   * @param {Message | Buffer} msg The message to be sent. For default AMQP format msg parameter
+   * should be of type Message interface. For a custom format, the msg parameter should be a Buffer
+   * and a valid value should be passed to the `format` argument.
+   * @param {Buffer | string} [tag] The message tag if any.
+   * @param {number} [format] The message format. Specify this if a message with custom format needs
+   * to be sent. `0` implies the standard AMQP 1.0 defined format. If no value is provided, then the
+   * given message is assumed to be of type Message interface and encoded appropriately.
+   * @returns {Promise<Delivery>} Promise<Delivery> The delivery information about the sent message.
+   */
+  send(msg: Message | Buffer, tag?: Buffer | string, format?: number): Promise<Delivery> {
+    return new Promise<Delivery>((resolve, reject) => {
+      if (this.sendable()) {
+        const timer = setTimeout(() => {
+          this.deliveryDispositionMap.delete(delivery.id);
+          const message = `[${this.connection.id}] Sender '${this.name}' on amqp session ` +
+            `'${this.session.id}', with address '${this.address}' was not able to send the ` +
+            `message with delivery id ${delivery.id} right now, due to operation timeout.`;
+          log.sender(message);
+          return reject(new OperationTimeoutError(message));
+        }, this.messageTimeoutInSeconds * 1000);
+
+        const delivery = (this._link as RheaSender).send(msg, tag, format);
+        this.deliveryDispositionMap.set(delivery.id, {
+          resolve: resolve,
+          reject: reject,
+          timer: timer
+        });
+      } else {
+        // Please send the message after some time.
+        const msg =
+          `[${this.connection.id}] Sender "${this.name}" on amqp session "${this.session.id}", ` +
+          `with address ${this.address} cannot send the message right now as it does not have ` +
+          `enough credit. Please try later.`;
+        log.error(msg);
+        reject(new Error(msg));
```

Comment:
This is one of those errors that users should be able to catch, identify and react accordingly. Using a normal `Error` with a custom message would force the user to take a dependency on the error message to decide the course of their action

Reply:
Yup. Have added two new Errors `InsufficentCreditError` and `SendOperationFailedError`